### ISSUE TITLE
bioconda integration

### ISSUE THREAD

**Issue by bgruening** [jhcepas/eggnog-mapper#28]
Is there any interest in creating a bioconda package and biocontainer for eggnog?
I could help in this process if you like.

**Comment by jhcepas** [comment]
It would be great! Please take the lead on this if you wish. 
I can help you with any problem. 
thanks

**Comment by jhcepas** [comment]
btw, @bgruening, I am trying to create bioconda recipes for missing ete3 external tools: phyml, paml, iqtree, t_coffee, slr, kalign, pmodeltest (https://github.com/jhcepas/bioconda-recipes/tree/phylotools). I've seen you just added paml. Are you working on any other?

**Comment by bgruening** [comment]
@jhcepas no I don't at the moment. But t_coffee should be available. Best way is probably to move this issue to bioconda and ask for help to be sure no work is redundant.

Thanks for working on it!

**Comment by caleb-easterly** [comment]
I have a recipe ready (attached), which was mostly written by @jj-umn. It does depend on the changes in PR #56, which is why I've pointed it to my fork. Thoughts? 
[eggnog-mapper.zip](https://github.com/jhcepas/eggnog-mapper/files/1296707/eggnog-mapper.zip)

**Comment by bgruening** [comment]
@caleb-easterly looks good to me! Can you submit this as a PR to bioconda?

**Comment by caleb-easterly** [comment]
Done! It's at https://github.com/bioconda/bioconda-recipes/pull/6008

**Comment by bgruening** [comment]
Thanks I commented on the Bioconda PR. This looks good! Do you plan to release a stable version soon?
Currently the Bioconda version do not fit to the beta version label of the tarball.

**Comment by jhcepas** [comment]
Yes, I think we are pretty ready to release version 1.0. I would like to add a couple of features that still live in separate branches though...
thanks for your work, btw

**Comment by bgruening** [comment]
No problem, I try to help where I can. Any ETA for a release. Should we have a eggnog beta Conda package, or do we want to wait until final is out?

**Comment by jhcepas** [comment]
I think we could have a conda pkg based on the latest release. The updated recipe should be pretty simple for next version, there are no changes in dependencies, etc...

**Comment by caleb-easterly** [comment]
Let me know if you want help with anything, @jhcepas. Did you have a chance to take a look at [egglet](https://github.com/caleb-easterly/egglet)? It's a functional mini version (5.6Mb) of the eggNOG database. In its current form, it would allow for testing one alignment, but it could be expanded to allow for more. I remember you saying that unittesting was still a TODO.

**Comment by caleb-easterly** [comment]
I cancelled the PR to bioconda-recipes - I can update the recipe once 1.0 is released. The main thing that it needs to work in conda (and Galaxy) is to have the ability to specify the data directory (#56, #44)

**Comment by epruesse** [comment]
@caleb-easterly Any progress on this?

**Comment by caleb-easterly** [comment]
As far as I understand, we're waiting on the next major release of eggnog-mapper before we distribute it on conda. @jhcepas, do you have an estimate for the next version? If it's a ways off it might be useful to package it for conda in the meantime, but up to you.

**Comment by epruesse** [comment]
Unless there are really drastic changes in the next release making the packaging a lot easier, or big problems with this release, I don't see a reason not to package, even if the next release might be soon.

**Comment by caleb-easterly** [comment]
I'm not sure what the changes are going to be, and if they'll be major or not. @jhcepas? 

I'm happy to re-open the PR for bioconda-recipes, if that's okay with everyone. My recipe was working, last I checked.

**Comment by jhcepas** [comment]
I am mainly working on two incremental changes (but moving slow, sorry): 
1. adding two new functional sources to the annotation database (KO and BiGG columns in the result file)
2. a mixed searching mode: first round using diamond, then use HMM for queries without diamond hits.

In any case, I think is safe to package current version and update the recipe when those addons are ready.

**Comment by caleb-easterly** [comment]
Great, I should have that done by tomorrow.

**Comment by epruesse** [comment]
ping me for review, could really use that package :)

**Comment by caleb-easterly** [comment]
Here's [the recipe](https://github.com/bioconda/bioconda-recipes/tree/eggnog-mapper/recipes/eggnog-mapper). I'm waiting to see if it passes Travis.  Nothing too wild, but note that I'm using my own fork as a source, because that allows for specifying a data directory separate from the install directory (this should allow for using the same eggnog database from within multiple conda environments).

**Comment by epruesse** [comment]
Will that feature (yes, very desirable) be merged upstream?

**Comment by epruesse** [comment]
For consistency, it would probably be better to use the upstream release archive and a patch. Also, I think using `setup.py` rather than `cp` would be advisable (although the setup.py may need patching as well to put the scripts in the correct places).

**Comment by jhcepas** [comment]
@caleb-easterly I can merge the PR that allows to specify data directory. 
is it this one or you have another one? https://github.com/jhcepas/eggnog-mapper/pull/44

**Comment by caleb-easterly** [comment]
The most recent one is #56 - now that I'm looking at it, though, I wonder if the `--dmnd_db` option is really needed. Ideally, a user would never pass their own diamond database, correct? It can be fixed so that the diamond database path is automatically defined as `data_dir/eggnog_proteins.dmnd`

**Comment by jhcepas** [comment]
@caleb-easterly I merged your changes, and I am very close to have the new functionality [ready](https://github.com/jhcepas/eggnog-mapper/pull/64) (KO and BiGG annotation sources). 

Only missing a decent unittest that checks the annotations produced for a few controlled queries. 
Do you have something in that direction?

**Comment by caleb-easterly** [comment]
Thanks, @jhcepas! 

I only have a very small test implemented in a Galaxy wrapper, based on the small eggnog database I have [here](https://github.com/caleb-easterly/egglet/tree/master/data). The query I've been using is a fragment of the Nlim_1033 sequence in the ENOG411CB2I orthologous group. With the current state of the database, only queries from the HMM of that OG group would produce hits. Would you want a slightly more expanded database to allow for different queries?

**Comment by jhcepas** [comment]
https://github.com/jhcepas/eggnog-mapper/releases/tag/1.0.0
This version integrates all changes and seems to be running ok for all web users...

**Comment by caleb-easterly** [comment]
Great! looking forward to working with the new version. 

I'm working on the [conda recipe](https://github.com/bioconda/bioconda-recipes/tree/eggnog-mapper/recipes/eggnog-mapper). @epruesse, would you be able to take a look at it? I switched to the new version, but I'm having trouble figuring out how to use setuptools with conda. The common command (`$PYTHON setup.py install`) isn't working - it's always installing into the wrong directories and I'm not sure how to patch `setup.py` (or change `build.sh`)

**Comment by epruesse** [comment]
@caleb-easterly Sorry, didn't catch that ping in time. Looks like you got it working though! 👍

**Comment by caleb-easterly** [comment]
Yes, the eggnog-mapper was merged into bioconda, and tests showed that it was working 
(PR: https://github.com/bioconda/bioconda-recipes/pull/6372). Should be good to go!

**Comment by caleb-easterly** [comment]
@jhcepas, how is the eggnog.db format changing? Are they upgrading the eggNOG database itself (i.e., from 4.5.1 to 4.5.2 or 4.6), or is the SQL database schema changing?

**Comment by jhcepas** [comment]
only the schema changes. Data remains de same (plus the new KO and BiGG tables).

`$sqlite3 eggnog.db`, then `.schema` should give you the new model. 

Regarding the version number for the emapper database: `4.5` refers to the eggnog data used, and I am planning to increment the last digit if I need to add more changes to the schema or add new tables.

**Comment by jhcepas** [comment]
btw, the update in release `1.0.1` is important. The script for downloading production databases had a typo.  Not sure how difficult is to update it in bioconda.
tx

**Comment by epruesse** [comment]
@caleb-easterly is already on it here: bioconda/bioconda-recipes#6497

**Comment by caleb-easterly** [comment]
Yes, I should have it merged by this afternoon. Except for large changes,
conda version updates should not be very difficult (provided one remembers
to change all the version numbers...)



Regards,

Caleb Easterly

Bioinformatics and Public Health Researcher
School of Public Health &
    Dept. of Biochemistry, Molecular Biology and Biophysics
University of Minnesota
7-166 MCB
420 Washington Avenue SE
Minneapolis, MN 55455

**Comment by epruesse** [comment]
You can use jinja templating to make a version upgrade a one-line patch.

**Comment by bgruening** [comment]
@jhcepas new package is online!
